In [1]:
import os
import sys


repo_dir = os.path.dirname(os.getcwd())


output_dir = os.path.join(repo_dir, "output")

sys.path.append(repo_dir)

In [2]:
from collections import Counter
import numpy as np
import pandas as pd
import plotly.graph_objects as go

from data_utils import read_cmu_mosei, create_dataset

/tmp/ipykernel_63375/3810907228.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


# 1. Load data

In [3]:
words, labels = read_cmu_mosei(
    data_dir="/home/vinicius/Documentos/Repositories/msc/data"
)

train, validation, test = create_dataset(words, labels)

[2024-03-04 18:24:46.688] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedPhones.csd ...
[2024-03-04 18:24:47.445] | Status  | Checking the integrity of the <phoneme> computational sequence ...
[2024-03-04 18:24:47.445] | Status  | Checking the format of the data in <phoneme> computational sequence ...


[2024-03-04 18:24:50.247] | Success | <phoneme> computational sequence data in correct format.
[2024-03-04 18:24:50.247] | Status  | Checking the format of the metadata in <phoneme> computational sequence ...
[2024-03-04 18:24:50.247] | Warning | <phoneme> computational sequence does not have all the required metadata ... continuing 
[2024-03-04 18:24:50.249] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedWordVectors.csd ...
[2024-03-04 18:24:50.912] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2024-03-04 18:24:50.912] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...


[2024-03-04 18:24:54.108] | Success | <glove_vectors> computational sequence data in correct format.
[2024-03-04 18:24:54.108] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2024-03-04 18:24:54.108] | Warning | <glove_vectors> computational sequence does not have all the required metadata ... continuing 
[2024-03-04 18:24:54.109] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_Labels.csd ...
[2024-03-04 18:24:54.584] | Status  | Checking the integrity of the <All Labels> computational sequence ...
[2024-03-04 18:24:54.584] | Status  | Checking the format of the data in <All Labels> computational sequence ...


[2024-03-04 18:24:56.057] | Success | <All Labels> computational sequence data in correct format.
[2024-03-04 18:24:56.057] | Status  | Checking the format of the metadata in <All Labels> computational sequence ...
[2024-03-04 18:24:56.057] | Warning | <All Labels> computational sequence does not have all the required metadata ... continuing 
[2024-03-04 18:24:56.059] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedWords.csd ...
[2024-03-04 18:24:56.616] | Status  | Checking the integrity of the <words> computational sequence ...
[2024-03-04 18:24:56.616] | Status  | Checking the format of the data in <words> computational sequence ...


[2024-03-04 18:24:58.138] | Success | <words> computational sequence data in correct format.
[2024-03-04 18:24:58.138] | Status  | Checking the format of the metadata in <words> computational sequence ...
[2024-03-04 18:24:58.138] | Warning | <words> computational sequence does not have all the required metadata ... continuing 
[2024-03-04 18:24:58.138] | Success | Dataset initialized successfully ... 


In [4]:
# Avaliação do tamanho do dataset
train_senteces = train.shape[0]
validation_sentences = validation.shape[0]
test_sentences = test.shape[0]
print(f"Número de sentenças de treinamento: {train_senteces}")
print(f"Número de sentenças do validação:   {validation_sentences}")
print(f"Número de sentenças do teste:       {test_sentences}")
print(
    f"Número total de sentenças:          {train_senteces + validation_sentences + test_sentences}"
)

df = pd.concat([train, validation, test], ignore_index=True)
display(df)

Número de sentenças de treinamento: 15807
Número de sentenças do validação:   1818
Número de sentenças do teste:       4512
Número total de sentenças:          22137


Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
2      antonio cassano won a a title during season wi...   0.333333   
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
...                                                  ...        ...   
22132  people in the world that understand and believ...   1.000000   
22133  the essence of addressing social justice from ...   0.000000   
22134  so adults that define the policies have to ens...   1.000000   
22135  when you fight for change when you fight for s...   0.666667   
22136  so it is not a question of being optimistic or...   0.666667   

       Happiness   Sadness  Anger      Fear  Disgust  Surprise  
0       0.333333  0.333333    0.0  0.000000      0.0  0.333333  
1       0.333333  0.000000    0.0  0.333333      0.0  0.000000  
2       0.333333  0.000000    0.0  0.000000      0.0  0.000000  
3       0.000000  0.000000    0.0  0.000000      0.0  0.000000  
4       0.000000  0.000000    0.0  0.000000      0.0  0.000000  
...          ...       ...    ...       ...      ...       ...  
22132   0.666667  0.000000    0.0  0.000000      0.0  0.000000  
22133   0.666667  0.000000    0.0  0.000000      0.0  0.000000  
22134   1.000000  0.333333    0.0  0.000000      0.0  0.333333  
22135   1.000000  0.333333    0.0  0.000000      0.0  0.000000  
22136   0.666667  0.000000    0.0  0.333333      0.0  0.000000  

[22137 rows x 8 columns]

In [5]:
df.loc[(df["Happiness"] > 1.5)]

Sentence  Sentiment  \
5      i'm samantha deutscher for innovations is one ...   0.333333   
40     hello my name's steve williams and a i'm full ...   1.000000   
42     stuff came out i highly recommend this air flo...   2.666667   
43                         so yeah that's my dental care   0.666667   
44              so that's it for today thank you guys so   2.333333   
...                                                  ...        ...   
22107  the case may be treat this as you would for an...   0.333333   
22111  i know about you but i am at my best when show...   1.666667   
22112                          may be thinking seriously   1.333333   
22113  the best part about doing an interview over th...   1.666667   
22121  julie canty and carlos a schmidt were recently...   0.333333   

       Happiness  Sadness  Anger  Fear  Disgust  Surprise  
5       1.666667      0.0    0.0   0.0      0.0       0.0  
40      1.666667      0.0    0.0   0.0      0.0       0.0  
42      1.666667      0.0    0.0   0.0      0.0       0.0  
43      2.000000      0.0    0.0   0.0      0.0       0.0  
44      2.000000      0.0    0.0   0.0      0.0       0.0  
...          ...      ...    ...   ...      ...       ...  
22107   1.666667      0.0    0.0   0.0      0.0       0.0  
22111   1.666667      0.0    0.0   0.0      0.0       0.0  
22112   2.000000      0.0    0.0   0.0      0.0       0.0  
22113   2.000000      0.0    0.0   0.0      0.0       0.0  
22121   1.666667      0.0    0.0   0.0      0.0       0.0  

[2122 rows x 8 columns]

# 2. Exploration Data Analysis

## 2.1 Neutral Sentences

In [6]:
# Decisão de classificar as sentenças como neutras
def neutral_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
    
    if all(row[emotion] == 0 for emotion in emotions):
        return "Neutral"
    else:
        return "Non-Neutral"


df["Emotion_Label"] = df.apply(neutral_label, axis=1)
df.loc[df["Emotion_Label"] == "Neutral"]

Sentence  Sentiment  \
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
21     say that i choose to become an attorney but i ...  -0.666667   
22     really have choice to go into that profession ...  -1.333333   
28     lends itself beautifully to entrepreneurial ac...   0.666667   
...                                                  ...        ...   
22063  do it right away don't wait until the problem ...   1.000000   
22064  crazy so it helps keep your resin doing what i...   1.666667   
22065  designed for well water because that's the kin...   0.333333   
22104                          until next time peace out   0.000000   
22116  caps and if that's the direction we're headed ...  -0.666667   

       Happiness  Sadness  Anger  Fear  Disgust  Surprise Emotion_Label  
3            0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
4            0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
21           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
28           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
...          ...      ...    ...   ...      ...       ...           ...  
22063        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22064        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22065        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22104        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22116        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  

[2974 rows x 9 columns]

In [7]:
# Existem frases classificadas sem nenhuma emocao?
neutral_phrases = df.loc[df["Emotion_Label"] == "Neutral"]

neutral_phrases.describe().loc[["min", "25%", "50%", "75%", "max"]]

Sentiment  Happiness  Sadness  Anger  Fear  Disgust  Surprise
min  -2.333333        0.0      0.0    0.0   0.0      0.0       0.0
25%   0.000000        0.0      0.0    0.0   0.0      0.0       0.0
50%   0.000000        0.0      0.0    0.0   0.0      0.0       0.0
75%   0.333333        0.0      0.0    0.0   0.0      0.0       0.0
max   2.666667        0.0      0.0    0.0   0.0      0.0       0.0

### 2.1.1 Percentage of Neutral Sentences in dataset

In [8]:
count_neutral_phrases = neutral_phrases.shape[0]

count_df = df.shape[0]

porcentage_neutral_phrases = np.round(count_neutral_phrases / count_df, 2) * 100


print(f"Porcentage of neutral phrases: {porcentage_neutral_phrases} %")

Porcentage of neutral phrases: 13.0 %


### 2.1.2 Quantity of neutral sentences per sentiment

In [9]:
count = neutral_phrases.groupby(by="Sentiment").count()["Sentence"]

percentage = neutral_phrases["Sentiment"].value_counts(normalize=True) * 100

summary = pd.DataFrame({"Count": count, "Percentage": percentage})

summary

Count  Percentage
Sentiment                   
-2.333333      1    0.033625
-2.000000      9    0.302623
-1.666667     38    1.277740
-1.333333     49    1.647613
-1.000000     57    1.916611
-0.666667    134    4.505716
-0.333333    247    8.305313
-0.166667      1    0.033625
 0.000000   1395   46.906523
 0.166667      1    0.033625
 0.333333    501   16.845999
 0.666667    285    9.583053
 1.000000    125    4.203093
 1.333333     75    2.521856
 1.500000      1    0.033625
 1.666667     33    1.109617
 2.000000     18    0.605245
 2.333333      3    0.100874
 2.666667      1    0.033625

### 2.1.3 Examples of Neutral Sentences with Positive Sentiment

In [10]:
positive_neutral_phrases = neutral_phrases.loc[neutral_phrases["Sentiment"] >= 2]

In [11]:
# Example of positive sentences
for _, s in positive_neutral_phrases.iterrows():
    sentence = s["Sentence"]
    width = 100
    print("-" * width)
    for i in range(0, len(sentence), width):
        print(sentence[i : i + width])
    print("-" * width)
    print()

----------------------------------------------------------------------------------------------------
president challenged us to be bold and dream big and that's exactly what we are doing at nasa while 
i was proud to fly on the space shuttle tomorrow's explorers will now dream of one day walking on ma
rs
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
one big option is to use artificial tears which are always very effective in helping your dry eyes s
ometimes make up can cause that and you want to refrain from wearing any makeup or anything like tha
t if you're getting dry eyes from it so to recap use a lot of fluids use aqua tears or any artificia
l tears substance and if you still continue to have lots of problems like that you may want to speak
 to an eye doctor thank you i'm dan carlson
-------------------------------------------

### 2.1.4 Examples of Neutral Sentences with Negative Sentiment

In [12]:
negative_neutral_phrases = neutral_phrases.loc[neutral_phrases["Sentiment"] <= -2]

In [13]:
# Example of negative sentences
for _, s in negative_neutral_phrases.iterrows():
    sentence = s["Sentence"]
    width = 100
    print("-" * width)
    for i in range(0, len(sentence), width):
        print(sentence[i : i + width])
    print("-" * width)
    print()

----------------------------------------------------------------------------------------------------
trump has refused to release his tax returns we know if he owes money to russian oil billionaires cl
ose to molly discussion in intelligence circles is that his behavior looks like someone who maybe co
mpromised who may be concerned about something and nobody knows what that is if financial ties or fi
nancial leverage something more than that i the only way for trump to deal with all these conflicts 
of interest is to sell his companies and then put someone besides his children in charge of the fina
ncial proceeds
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
to give this a one of five star don't worry
----------------------------------------------------------------------------------------------------

------------------------------

## 2.2 Sentiment to Label
Estrategia de arredondar a média dos anotadores e transformar em variável categórica

In [14]:
def sentiment_to_category(row):
    sentiment = {
        -3: "highly negative",
        -2: "negative",
        -1: "weakly negative",
        0: "neutral",
        1: "weakly positive",
        2: "positive",
        3: "highly positive",
    }

    numerical_sentiment = row
    numerical_sentiment_rounded = np.round(numerical_sentiment)
    return sentiment.get(numerical_sentiment_rounded)


df["Sentiment_Categorical"] = df["Sentiment"].apply(sentiment_to_category)
display(df)

Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
2      antonio cassano won a a title during season wi...   0.333333   
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
...                                                  ...        ...   
22132  people in the world that understand and believ...   1.000000   
22133  the essence of addressing social justice from ...   0.000000   
22134  so adults that define the policies have to ens...   1.000000   
22135  when you fight for change when you fight for s...   0.666667   
22136  so it is not a question of being optimistic or...   0.666667   

       Happiness   Sadness  Anger      Fear  Disgust  Surprise Emotion_Label  \
0       0.333333  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
1       0.333333  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   
2       0.333333  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
3       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
4       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
...          ...       ...    ...       ...      ...       ...           ...   
22132   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22133   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22134   1.000000  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
22135   1.000000  0.333333    0.0  0.000000      0.0  0.000000   Non-Neutral   
22136   0.666667  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   

      Sentiment_Categorical  
0                   neutral  
1           weakly positive  
2                   neutral  
3                   neutral  
4           weakly negative  
...                     ...  
22132       weakly positive  
22133               neutral  
22134       weakly positive  
22135       weakly positive  
22136       weakly positive  

[22137 rows x 10 columns]

In [15]:
df.groupby(by=["Sentiment", "Sentiment_Categorical"])["Sentence"].count().reset_index()

Sentiment Sentiment_Categorical  Sentence
0   -3.000000       highly negative       124
1   -2.666667       highly negative       278
2   -2.333333              negative       409
3   -2.000000              negative       644
4   -1.666667              negative       752
5   -1.333333       weakly negative       823
6   -1.000000       weakly negative       887
7   -0.666667       weakly negative      1106
8   -0.500000               neutral         4
9   -0.333333               neutral      1513
10  -0.166667               neutral         3
11   0.000000               neutral      4542
12   0.166667               neutral        11
13   0.333333               neutral      2933
14   0.500000               neutral        11
15   0.666667       weakly positive      2460
16   0.833333       weakly positive         7
17   1.000000       weakly positive      1850
18   1.166667       weakly positive         3
19   1.333333       weakly positive      1389
20   1.500000              positive         8
21   1.666667              positive       994
22   1.833333              positive         2
23   2.000000              positive       728
24   2.333333              positive       393
25   2.500000              positive         1
26   2.666667       highly positive       202
27   3.000000       highly positive        60

In [16]:
neutral_phrases["Sentiment_Categorical"] = df["Sentiment"].apply(sentiment_to_category)

/tmp/ipykernel_63375/4071804904.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  neutral_phrases["Sentiment_Categorical"] = df["Sentiment"].apply(sentiment_to_category)


In [17]:
count = neutral_phrases.groupby(by="Sentiment_Categorical").count()["Sentence"]

percentage = neutral_phrases["Sentiment_Categorical"].value_counts(normalize=True) * 100

summary = pd.DataFrame({"Count": count, "Percentage": percentage})

summary

Count  Percentage
Sentiment_Categorical                   
highly positive            1    0.033625
negative                  48    1.613988
neutral                 2145   72.125084
positive                  55    1.849361
weakly negative          240    8.069939
weakly positive          485   16.308003

## 2.3 Emotions

Este conjunto de dados contém múltiplas sentenças, cada uma classificada com base na média de avaliações fornecidas por três anotadores diferentes, utilizando uma escala Likert.

Escala Likert:
 - 0: Nenhuma evidência da emoção.
 - 1: Pouca evidência da emoção.
 - 2: Evidência da emoção.
 - 3: Forte evidência da emoção.

Cada sentença é representada por um vetor numérico, onde cada elemento corresponde a uma das seis emoções identificadas:

Exemplo de vetor de emoções:

Happiness
Sadness
Anger
Fear
Disgust
Surprise
Exemplo de sentença rotulada:

text = [0.0, 0.3333, 0.0, 1.6667, 0.0, 0.0]

Classificação dos Rótulos:
A tarefa agora é conceber uma metodologia para converter estes rótulos numéricos em uma única categoria representativa para cada frase. As possíveis categorias são:

 - UniqueLabel: Sentenças que contêm apenas uma única média no vetor.
 - MultiLabel: Sentenças que contêm duas ou mais médias diferentes no vetor.
 - EqualMultiLabel: Sentenças que contêm duas ou mais médias iguais no vetor.

A classificação adequada é crucial para análises subsequentes e para treinar modelos de machine learning que dependem de rótulos claros e distintos.

In [18]:
def multiple_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
    multiple_label = [emotion for emotion in emotions if row[emotion] > 0]
    if len(multiple_label) > 1:
        return True
    return False


def multiple_label_equal_average(row):
    """
    Verifica se uma linha possui múltiplos rótulos com a mesma média.

    Parameters:
    row (dict): Uma linha do dataset, onde as chaves são rótulos de emoções e os valores são suas respectivas médias.

    Returns:
    bool: Retorna True se existirem múltiplos rótulos com a mesma média máxima, caso contrário, False.
    """

    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Extraindo as médias (ignorando rótulos com média 0)
    averages = [row[emotion] for emotion in emotions if row[emotion] > 0]

    # Verifica se existem múltiplos rótulos e se a maior média aparece mais de uma vez
    if len(averages) > 1 and averages.count(max(averages)) > 1:
        return True
    return False


df["Multiple_Label"] = df.apply(multiple_label, axis=1)
df["Equal_Average"] = df.apply(multiple_label_equal_average, axis=1)

multiple_label_df = df.loc[df["Multiple_Label"] == True]
unique_label_df = df.loc[df["Multiple_Label"] == False]
equal_average_df = df.loc[df["Equal_Average"] == True]

count_unique_label = unique_label_df.shape[0]
count_multiple_label = multiple_label_df.shape[0]
count_equal_average = unique_label_df.shape[0]
count_non_equal_average_multiple_labels = count_multiple_label - count_equal_average

display(df)

Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
2      antonio cassano won a a title during season wi...   0.333333   
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
...                                                  ...        ...   
22132  people in the world that understand and believ...   1.000000   
22133  the essence of addressing social justice from ...   0.000000   
22134  so adults that define the policies have to ens...   1.000000   
22135  when you fight for change when you fight for s...   0.666667   
22136  so it is not a question of being optimistic or...   0.666667   

       Happiness   Sadness  Anger      Fear  Disgust  Surprise Emotion_Label  \
0       0.333333  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
1       0.333333  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   
2       0.333333  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
3       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
4       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
...          ...       ...    ...       ...      ...       ...           ...   
22132   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22133   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22134   1.000000  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
22135   1.000000  0.333333    0.0  0.000000      0.0  0.000000   Non-Neutral   
22136   0.666667  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   

      Sentiment_Categorical  Multiple_Label  Equal_Average  
0                   neutral            True           True  
1           weakly positive            True           True  
2                   neutral           False          False  
3                   neutral           False          False  
4           weakly negative           False          False  
...                     ...             ...            ...  
22132       weakly positive           False          False  
22133               neutral           False          False  
22134       weakly positive            True          False  
22135       weakly positive            True          False  
22136       weakly positive            True          False  

[22137 rows x 12 columns]

### 2.3.1 UniqueLabel

Example Unique Label

In [19]:
display(unique_label_df)

Sentence  Sentiment  \
2      antonio cassano won a a title during season wi...   0.333333   
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
5      i'm samantha deutscher for innovations is one ...   0.333333   
7      innovations is focussed on the acquisition dev...   1.000000   
...                                                  ...        ...   
22127  that productivity and competitiveness leads to...   0.333333   
22128  you have a fair chance at whatever wealth you ...   1.666667   
22129  that the ilo has highlighted that there is a l...   1.000000   
22132  people in the world that understand and believ...   1.000000   
22133  the essence of addressing social justice from ...   0.000000   

       Happiness  Sadness  Anger  Fear  Disgust  Surprise Emotion_Label  \
2       0.333333      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
3       0.000000      0.0    0.0   0.0      0.0       0.0       Neutral   
4       0.000000      0.0    0.0   0.0      0.0       0.0       Neutral   
5       1.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
7       0.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
...          ...      ...    ...   ...      ...       ...           ...   
22127   0.333333      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
22128   0.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
22129   0.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
22132   0.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   
22133   0.666667      0.0    0.0   0.0      0.0       0.0   Non-Neutral   

      Sentiment_Categorical  Multiple_Label  Equal_Average  
2                   neutral           False          False  
3                   neutral           False          False  
4           weakly negative           False          False  
5                   neutral           False          False  
7           weakly positive           False          False  
...                     ...             ...            ...  
22127               neutral           False          False  
22128              positive           False          False  
22129       weakly positive           False          False  
22132       weakly positive           False          False  
22133               neutral           False          False  

[13838 rows x 12 columns]

Para sentenças que apresentam um único valor de média em seu vetor, a estratégia adotada será a de rotulá-las de acordo com a categoria correspondente a essa média específica.

### 2.3.2 MultiLabel

In [20]:
def multiple_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
    multiple_label = [emotion for emotion in emotions if row[emotion] > 0]
    if len(multiple_label) > 1:
        return True
    return False


def multiple_label_equal_average(row):
    """
    Verifica se uma linha possui múltiplos rótulos com a mesma média.

    Parameters:
    row (dict): Uma linha do dataset, onde as chaves são rótulos de emoções e os valores são suas respectivas médias.

    Returns:
    bool: Retorna True se existirem múltiplos rótulos com a mesma média máxima, caso contrário, False.
    """

    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Extraindo as médias (ignorando rótulos com média 0)
    averages = [row[emotion] for emotion in emotions if row[emotion] > 0]

    # Verifica se existem múltiplos rótulos e se a maior média aparece mais de uma vez
    if len(averages) > 1 and averages.count(max(averages)) > 1:
        return True
    return False


df["Multiple_Label"] = df.apply(multiple_label, axis=1)
df["Equal_Average"] = df.apply(multiple_label_equal_average, axis=1)

multiple_label_df = df.loc[df["Multiple_Label"] == True].copy()
unique_label_df = df.loc[df["Multiple_Label"] == False].copy()
equal_average_df = df.loc[df["Equal_Average"] == True].copy()

count_unique_label = unique_label_df.shape[0]
count_multiple_label = multiple_label_df.shape[0]
count_equal_average = equal_average_df.shape[0]
count_non_equal_average_multiple_labels = count_multiple_label - count_equal_average

#### Example MultiLabel

In [21]:
display(multiple_label_df)

Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
6      with its new listing the company changed its n...  -1.000000   
8      company has purchased all rights intellectual ...   0.666667   
16     same and a mistake to try to separate them out...  -1.000000   
...                                                  ...        ...   
22130  key question is that we have to move from a gl...   0.333333   
22131  has made it more difficult and why the struggl...  -0.333333   
22134  so adults that define the policies have to ens...   1.000000   
22135  when you fight for change when you fight for s...   0.666667   
22136  so it is not a question of being optimistic or...   0.666667   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.333333  0.333333  0.000000  0.000000  0.000000  0.333333   
1       0.333333  0.000000  0.000000  0.333333  0.000000  0.000000   
6       0.333333  0.000000  0.000000  0.000000  0.333333  0.000000   
8       0.666667  0.000000  0.666667  0.333333  0.000000  0.000000   
16      0.000000  0.333333  0.666667  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22130   0.666667  0.666667  0.000000  0.000000  0.000000  0.000000   
22131   0.333333  0.333333  0.000000  0.000000  0.000000  0.000000   
22134   1.000000  0.333333  0.000000  0.000000  0.000000  0.333333   
22135   1.000000  0.333333  0.000000  0.000000  0.000000  0.000000   
22136   0.666667  0.000000  0.000000  0.333333  0.000000  0.000000   

      Emotion_Label Sentiment_Categorical  Multiple_Label  Equal_Average  
0       Non-Neutral               neutral            True           True  
1       Non-Neutral       weakly positive            True           True  
6       Non-Neutral       weakly negative            True           True  
8       Non-Neutral       weakly positive            True           True  
16      Non-Neutral       weakly negative            True          False  
...             ...                   ...             ...            ...  
22130   Non-Neutral               neutral            True           True  
22131   Non-Neutral               neutral            True           True  
22134   Non-Neutral       weakly positive            True          False  
22135   Non-Neutral       weakly positive            True          False  
22136   Non-Neutral       weakly positive            True          False  

[8299 rows x 12 columns]

#### Example EqualMultiLabel

In [22]:
display(equal_average_df)

Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
6      with its new listing the company changed its n...  -1.000000   
8      company has purchased all rights intellectual ...   0.666667   
27              form of a pixar movie and that structure   0.333333   
...                                                  ...        ...   
22086  really dull story line it doesn't make a lot o...  -2.333333   
22095             uhh it's just it's far too long if you  -1.333333   
22100  so uhh i give the movie two out of five stars ...  -1.666667   
22130  key question is that we have to move from a gl...   0.333333   
22131  has made it more difficult and why the struggl...  -0.333333   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.333333  0.333333  0.000000  0.000000  0.000000  0.333333   
1       0.333333  0.000000  0.000000  0.333333  0.000000  0.000000   
6       0.333333  0.000000  0.000000  0.000000  0.333333  0.000000   
8       0.666667  0.000000  0.666667  0.333333  0.000000  0.000000   
27      0.333333  0.333333  0.000000  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22086   0.000000  0.333333  0.000000  0.000000  0.333333  0.000000   
22095   0.000000  0.666667  0.666667  0.000000  0.333333  0.000000   
22100   0.333333  0.000000  0.000000  0.000000  0.333333  0.000000   
22130   0.666667  0.666667  0.000000  0.000000  0.000000  0.000000   
22131   0.333333  0.333333  0.000000  0.000000  0.000000  0.000000   

      Emotion_Label Sentiment_Categorical  Multiple_Label  Equal_Average  
0       Non-Neutral               neutral            True           True  
1       Non-Neutral       weakly positive            True           True  
6       Non-Neutral       weakly negative            True           True  
8       Non-Neutral       weakly positive            True           True  
27      Non-Neutral               neutral            True           True  
...             ...                   ...             ...            ...  
22086   Non-Neutral              negative            True           True  
22095   Non-Neutral       weakly negative            True           True  
22100   Non-Neutral              negative            True           True  
22130   Non-Neutral               neutral            True           True  
22131   Non-Neutral               neutral            True           True  

[2810 rows x 12 columns]

#### Quantificando

In [23]:
print(f"Dataset            -> {count_df}")
print(f"UniqueLabel        -> {count_unique_label}")
print(f"MultiLabel         -> {count_multiple_label}")
print(f"Percentage         -> {np.round(count_multiple_label / count_df, 2) * 100}%")
print()

print(f"MultiLabel         -> {count_multiple_label}")
print(f"NonEqualMultiLabel -> {count_non_equal_average_multiple_labels}")
print(f"EqualMultiLabel    -> {count_equal_average}")

print()


print(
    f"Percentage of Multiple Labels com médias iguais {np.round( count_equal_average / count_multiple_label , 2) * 100}%"
)
print(
    f"Percentual Multiple Labels sem médias iguais {np.round(count_non_equal_average_multiple_labels / count_multiple_label, 2) * 100 }%"
)
print()

print("Porcentagem do dataset", count_equal_average / count_df)
print(f"Porcentam do dataset {count_non_equal_average_multiple_labels / count_df}")

Dataset            -> 22137
UniqueLabel        -> 13838
MultiLabel         -> 8299
Percentage         -> 37.0%

MultiLabel         -> 8299
NonEqualMultiLabel -> 5489
EqualMultiLabel    -> 2810

Percentage of Multiple Labels com médias iguais 34.0%
Percentual Multiple Labels sem médias iguais 66.0%

Porcentagem do dataset 0.12693680263811719
Porcentam do dataset 0.24795591091837196


#### 2.3.2.1 EqualMultiLabel

Distribuition of Repetitive Averages

In [24]:
def count_non_zero_equal_averages(row):
    # Filtra as médias maiores que zero e conta quantas vezes cada média aparece
    return pd.Series(row[row > 0].values).value_counts().max()


# Aplicando a função em cada linha para contar o número máximo de médias iguais
equal_average_df["Max_Equal_Averages"] = equal_average_df[
    ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
].apply(count_non_zero_equal_averages, axis=1)

# Agrupando pelo número de máximas médias iguais e contando as sentenças
equal_average_count = (
    equal_average_df.groupby("Max_Equal_Averages").count()["Sentence"].reset_index()
)

# Renomeando as colunas para maior clareza
equal_average_count.columns = ["Number of Equal Averages", "Sentence Count"]

# Exibindo o resultado
display(equal_average_count)

Number of Equal Averages  Sentence Count
0                         2            2337
1                         3             404
2                         4              59
3                         5              10

Distribuition of Average

In [25]:
def get_combinations(row):
    # Filtra as médias maiores que zero
    non_zero_averages = row[row > 0]
    # Conta a frequência de cada média não nula
    counts = non_zero_averages.value_counts()

    # Se todas as médias são únicas, não haverá combinações
    if all(value == 1 for value in counts):
        return None

    # Encontra a maior média entre as mais comuns
    max_count = counts.max()
    most_common_averages = counts[counts == max_count].index
    most_common_average = max(most_common_averages)

    return most_common_average, max_count


# Aplicando a função em cada linha para obter a média mais comum e sua contagem
combination_results = equal_average_df[
    ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
].apply(get_combinations, axis=1)

# Contando a frequência de cada combinação
combination_counts = {}

for result in combination_results:
    if result is not None:
        average, count = result
        if (average, count) not in combination_counts:
            combination_counts[(average, count)] = 1
        else:
            combination_counts[(average, count)] += 1

# Convertendo o dicionário em uma lista de dicionários para criar um DataFrame
combination_data = [
    {"Average": average, "Combination Size": size, "Count": count}
    for (average, size), count in combination_counts.items()
]

# Convertendo em DataFrame
combination_df = pd.DataFrame(combination_data)

# Exibindo os resultados
display(combination_df.groupby(by=["Average", "Combination Size"]).sum())

Count
Average  Combination Size       
0.166667 2                     4
0.333333 2                  1331
         3                   314
         4                    52
         5                    10
0.666667 2                   719
         3                    80
         4                     7
1.000000 2                   214
         3                     8
1.333333 2                    38
1.666667 2                    21
         3                     1
2.000000 2                     7
         3                     1
2.333333 2                     2
2.666667 2                     1

# 3. Data cleaning

In [26]:
def emotions_to_category(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Criar uma lista de tuplas (emocao, média)
    emotions_average = [(emotion, row[emotion]) for emotion in emotions]

    # Verifica se todos as médias são zeros
    if all(average == 0 for _, average in emotions_average):
        return "Neutral"

    # Conta quantas tuplas têm a média diferente de zero
    count_non_zero = sum(average != 0 for _, average in emotions_average)

    # Verifica se apenas uma tupla têm média diferente de zero
    if count_non_zero == 1:
        for emotion, average in emotions_average:
            if average != 0:
                return emotion
    
    # MultiLabel: Escolhe a emoção com a maior média se houver mais de uma média diferente de zero
    if count_non_zero > 1:
        sorted_emotions = sorted(emotions_average, key=lambda x: x[1], reverse=True)
        # Verifica se há um empate na maior média
        if sorted_emotions[0][1] == sorted_emotions[1][1]:
            return "Indefinido"  # Ou pode retornar None
        return sorted_emotions[0][0]

    return None

df["Emotion_Category"] = df.apply(emotions_to_category, axis=1)
display(df)

Sentence  Sentiment  \
0      patricia sellers hi my name is patricia seller...   0.000000   
1      antonio cassano became most expensive teenager...   0.666667   
2      antonio cassano won a a title during season wi...   0.333333   
3      antonio cassano in action against john terry d...   0.000000   
4      actually tried to sign the old in february but...  -0.666667   
...                                                  ...        ...   
22132  people in the world that understand and believ...   1.000000   
22133  the essence of addressing social justice from ...   0.000000   
22134  so adults that define the policies have to ens...   1.000000   
22135  when you fight for change when you fight for s...   0.666667   
22136  so it is not a question of being optimistic or...   0.666667   

       Happiness   Sadness  Anger      Fear  Disgust  Surprise Emotion_Label  \
0       0.333333  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
1       0.333333  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   
2       0.333333  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
3       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
4       0.000000  0.000000    0.0  0.000000      0.0  0.000000       Neutral   
...          ...       ...    ...       ...      ...       ...           ...   
22132   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22133   0.666667  0.000000    0.0  0.000000      0.0  0.000000   Non-Neutral   
22134   1.000000  0.333333    0.0  0.000000      0.0  0.333333   Non-Neutral   
22135   1.000000  0.333333    0.0  0.000000      0.0  0.000000   Non-Neutral   
22136   0.666667  0.000000    0.0  0.333333      0.0  0.000000   Non-Neutral   

      Sentiment_Categorical  Multiple_Label  Equal_Average Emotion_Category  
0                   neutral            True           True       Indefinido  
1           weakly positive            True           True       Indefinido  
2                   neutral           False          False        Happiness  
3                   neutral           False          False          Neutral  
4           weakly negative           False          False          Neutral  
...                     ...             ...            ...              ...  
22132       weakly positive           False          False        Happiness  
22133               neutral           False          False        Happiness  
22134       weakly positive            True          False        Happiness  
22135       weakly positive            True          False        Happiness  
22136       weakly positive            True          False        Happiness  

[22137 rows x 13 columns]

In [27]:
cleaned_df = df.loc[:, ["Sentence", "Emotion_Category"]].copy()
display(cleaned_df)

Sentence Emotion_Category
0      patricia sellers hi my name is patricia seller...       Indefinido
1      antonio cassano became most expensive teenager...       Indefinido
2      antonio cassano won a a title during season wi...        Happiness
3      antonio cassano in action against john terry d...          Neutral
4      actually tried to sign the old in february but...          Neutral
...                                                  ...              ...
22132  people in the world that understand and believ...        Happiness
22133  the essence of addressing social justice from ...        Happiness
22134  so adults that define the policies have to ens...        Happiness
22135  when you fight for change when you fight for s...        Happiness
22136  so it is not a question of being optimistic or...        Happiness

[22137 rows x 2 columns]

In [28]:
cleaned_df.loc[cleaned_df["Emotion_Category"] != "Indefinido"]

Sentence Emotion_Category
2      antonio cassano won a a title during season wi...        Happiness
3      antonio cassano in action against john terry d...          Neutral
4      actually tried to sign the old in february but...          Neutral
5      i'm samantha deutscher for innovations is one ...        Happiness
7      innovations is focussed on the acquisition dev...        Happiness
...                                                  ...              ...
22132  people in the world that understand and believ...        Happiness
22133  the essence of addressing social justice from ...        Happiness
22134  so adults that define the policies have to ens...        Happiness
22135  when you fight for change when you fight for s...        Happiness
22136  so it is not a question of being optimistic or...        Happiness

[19327 rows x 2 columns]

In [54]:
cleaned_df = cleaned_df.loc[cleaned_df["Emotion_Category"] != "Indefinido"].copy()


train_sentences = train["Sentence"].values
test_sentences = test["Sentence"].values
validation_sentences = validation["Sentence"].values

# Agrupar por 'Sentence' e 'Emotion_Category', contando ocorrências
display("Treinamento: ")
display(cleaned_df.loc[cleaned_df["Sentence"].isin(train_sentences)]["Emotion_Category"].value_counts())
print()

display("Validação")
display(cleaned_df.loc[cleaned_df["Sentence"].isin(validation_sentences)]["Emotion_Category"].value_counts())
print()

display("Teste")
display(cleaned_df.loc[cleaned_df["Sentence"].isin(test_sentences)]["Emotion_Category"].value_counts())
print()
# Se você tem DataFrames separados para train, test, validation, você faria:
# train_counts = train.groupby(['Sentence', 'Emotion_Category']).size().reset_index(name='Counts')
# test_counts = test.groupby(['Sentence', 'Emotion_Category']).size().reset_index(name='Counts')
# validation_counts = validation.groupby(['Sentence', 'Emotion_Category']).size().reset_index(name='Counts')





'Treinamento: '

Emotion_Category
Happiness    6673
Neutral      2089
Sadness      1972
Anger        1567
Disgust       931
Surprise      310
Fear          248
Name: count, dtype: int64

'Validação'

Emotion_Category
Happiness    747
Neutral      275
Sadness      274
Anger        155
Disgust       88
Surprise      40
Fear          26
Name: count, dtype: int64

'Teste'

Emotion_Category
Happiness    1918
Neutral       624
Sadness       539
Anger         483
Disgust       248
Surprise       99
Fear           80
Name: count, dtype: int64

In [50]:
validation.shape

(1818, 8)

In [34]:
cleaned_df.loc[cleaned_df["Emotion_Category"] != "Indefinido"]["Emotion_Category"].value_counts()

Emotion_Category
Happiness    9313
Neutral      2974
Sadness      2778
Anger        2200
Disgust      1262
Surprise      447
Fear          353
Name: count, dtype: int64

In [43]:
validation["Sentence"].values

array(['and you try to help them',
       'to write an angry letter of protest is not as effective as to write a love letter',
       'because when you are able to write a love letter you touch the best thing in the politician in the leader',
       ...,
       "okay we have talked about the person and the family in this birthday speech we're going to talk more about the person but right now what we want to do is talk about them in relation to their work and their work achievements what is it that they do are they a great secretary are they an amazing doctor are they a wonderful veterinarian are they a filmmaker are they an",
       'actor whatever it is you want to take this time in the speech to talk about what it is they do specifically any achievements that they have in relation to their work any awards',
       "that they've won or raises or promotions or rushing up the corporate ladder whatever it is you want to take this time and you also might want to mention what it is did thi